// @vexip-ui/hooks/dist/index.mjs
import { computed as W, watch as j, unref as E, getCurrentScope as Re, onScopeDispose as U, ref as C, onMounted as ct, nextTick as Vt, onBeforeUnmount as at, renderSlot as yn, isVNode as xn, Comment as En, createTextVNode as Rn, Fragment as Sn, toRef as On, readonly as Xt, isRef as ae, customRef as Tn, reactive as Je, shallowRef as Ve, watchEffect as gt } from "vue";
const F = typeof window < "u";
var bt;
F && ((bt = window == null ? void 0 : window.navigator) != null && bt.userAgent) && /iP(ad|hone|od)/.test(window.navigator.userAgent);
function Be(e) {
  return e != null;
}
function B() {
}
const zn = Object.freeze({
  aliceblue: "f0f8ff",
  antiquewhite: "faebd7",
  aqua: "0ff",
  aquamarine: "7fffd4",
  azure: "f0ffff",
  beige: "f5f5dc",
  bisque: "ffe4c4",
  black: "000",
  blanchedalmond: "ffebcd",
  blue: "00f",
  blueviolet: "8a2be2",
  brown: "a52a2a",
  burlywood: "deb887",
  burntsienna: "ea7e5d",
  cadetblue: "5f9ea0",
  chartreuse: "7fff00",
  chocolate: "d2691e",
  coral: "ff7f50",
  cornflowerblue: "6495ed",
  cornsilk: "fff8dc",
  crimson: "dc143c",
  cyan: "0ff",
  darkblue: "00008b",
  darkcyan: "008b8b",
  darkgoldenrod: "b8860b",
  darkgray: "a9a9a9",
  darkgreen: "006400",
  darkgrey: "a9a9a9",
  darkkhaki: "bdb76b",
  darkmagenta: "8b008b",
  darkolivegreen: "556b2f",
  darkorange: "ff8c00",
  darkorchid: "9932cc",
  darkred: "8b0000",
  darksalmon: "e9967a",
  darkseagreen: "8fbc8f",
  darkslateblue: "483d8b",
  darkslategray: "2f4f4f",
  darkslategrey: "2f4f4f",
  darkturquoise: "00ced1",
  darkviolet: "9400d3",
  deeppink: "ff1493",
  deepskyblue: "00bfff",
  dimgray: "696969",
  dimgrey: "696969",
  dodgerblue: "1e90ff",
  firebrick: "b22222",
  floralwhite: "fffaf0",
  forestgreen: "228b22",
  fuchsia: "f0f",
  gainsboro: "dcdcdc",
  ghostwhite: "f8f8ff",
  gold: "ffd700",
  goldenrod: "daa520",
  gray: "808080",
  green: "008000",
  greenyellow: "adff2f",
  grey: "808080",
  honeydew: "f0fff0",
  hotpink: "ff69b4",
  indianred: "cd5c5c",
  indigo: "4b0082",
  ivory: "fffff0",
  khaki: "f0e68c",
  lavender: "e6e6fa",
  lavenderblush: "fff0f5",
  lawngreen: "7cfc00",
  lemonchiffon: "fffacd",
  lightblue: "add8e6",
  lightcoral: "f08080",
  lightcyan: "e0ffff",
  lightgoldenrodyellow: "fafad2",
  lightgray: "d3d3d3",
  lightgreen: "90ee90",
  lightgrey: "d3d3d3",
  lightpink: "ffb6c1",
  lightsalmon: "ffa07a",
  lightseagreen: "20b2aa",
  lightskyblue: "87cefa",
  lightslategray: "789",
  lightslategrey: "789",
  lightsteelblue: "b0c4de",
  lightyellow: "ffffe0",
  lime: "0f0",
  limegreen: "32cd32",
  linen: "faf0e6",
  magenta: "f0f",
  maroon: "800000",
  mediumaquamarine: "66cdaa",
  mediumblue: "0000cd",
  mediumorchid: "ba55d3",
  mediumpurple: "9370db",
  mediumseagreen: "3cb371",
  mediumslateblue: "7b68ee",
  mediumspringgreen: "00fa9a",
  mediumturquoise: "48d1cc",
  mediumvioletred: "c71585",
  midnightblue: "191970",
  mintcream: "f5fffa",
  mistyrose: "ffe4e1",
  moccasin: "ffe4b5",
  navajowhite: "ffdead",
  navy: "000080",
  oldlace: "fdf5e6",
  olive: "808000",
  olivedrab: "6b8e23",
  orange: "ffa500",
  orangered: "ff4500",
  orchid: "da70d6",
  palegoldenrod: "eee8aa",
  palegreen: "98fb98",
  paleturquoise: "afeeee",
  palevioletred: "db7093",
  papayawhip: "ffefd5",
  peachpuff: "ffdab9",
  peru: "cd853f",
  pink: "ffc0cb",
  plum: "dda0dd",
  powderblue: "b0e0e6",
  purple: "800080",
  rebeccapurple: "663399",
  red: "f00",
  rosybrown: "bc8f8f",
  royalblue: "4169e1",
  saddlebrown: "8b4513",
  salmon: "fa8072",
  sandybrown: "f4a460",
  seagreen: "2e8b57",
  seashell: "fff5ee",
  sienna: "a0522d",
  silver: "c0c0c0",
  skyblue: "87ceeb",
  slateblue: "6a5acd",
  slategray: "708090",
  slategrey: "708090",
  snow: "fffafa",
  springgreen: "00ff7f",
  steelblue: "4682b4",
  tan: "d2b48c",
  teal: "008080",
  thistle: "d8bfd8",
  tomato: "ff6347",
  turquoise: "40e0d0",
  violet: "ee82ee",
  wheat: "f5deb3",
  white: "fff",
  whitesmoke: "f5f5f5",
  yellow: "ff0",
  yellowgreen: "9acd32"
});
Object.freeze(new Set(Object.keys(zn)));
function wt(e) {
  return e & -e;
}
function Cn(e, t = 0) {
  const n = new Array(e + 1).fill(0);
  function r(s, a) {
    if (!(!a || s >= e))
      for (s += 1; s <= e; )
        n[s] += a, s += wt(s);
  }
  function o(s = e) {
    if (s <= 0) return 0;
    s > e && (s = e);
    let a = s * t;
    for (; s > 0; )
      a += n[s], s -= wt(s);
    return a;
  }
  function i(s) {
    return o(s + 1) - o(s);
  }
  function c(s) {
    let a = 0, f = e;
    for (; f > a; ) {
      const l = Math.floor((a + f) / 2), u = o(l);
      if (u > s) {
        f = l;
        continue;
      } else if (u < s) {
        if (a === l)
          return o(a + 1) <= s ? a + 1 : a;
        a = l;
      } else
        return l;
    }
    return a;
  }
  return { tree: n, add: r, sum: o, get: i, boundIndex: c };
}
function kn(e) {
  let t = parseFloat(e);
  return Number.isNaN(t) && (t = Number(e)), Number.isNaN(t) ? 0 : t;
}
const An = F && ("ontouchstart" in window || Mn() > 0), yt = An ? "pointerdown" : "click";
function Mn() {
  return typeof navigator < "u" && (navigator.maxTouchPoints || navigator.msMaxTouchPoints) || 0;
}
function Ln(e, t, n = window.Event) {
  const { type: r, bubbles: o = !1, cancelable: i = !1, ...c } = t;
  if (!Be(r) || r === "") return !1;
  let s;
  return Be(n) ? s = new n(r, { bubbles: o, cancelable: i }) : (s = document.createEvent("HTMLEvents"), s.initEvent(r, o, i)), Object.assign(s, c), e.dispatchEvent(s);
}
const Pn = F ? requestAnimationFrame : (e) => {
  setTimeout(e, 16);
};
function jt(e, t = 16) {
  if (typeof e != "function")
    return B;
  const n = (...i) => {
    e(...i);
  };
  if (t <= 0)
    return Bn(n);
  let r = 0, o;
  return function(...i) {
    const c = Date.now(), s = c - r;
    clearTimeout(o), s >= t ? (r = c, n(...i)) : o = setTimeout(
      () => {
        r = Date.now(), n(...i);
      },
      Math.max(0, t - s)
    );
  };
}
function Bn(e) {
  if (typeof e != "function")
    return e;
  let t = !1, n, r;
  return function(...o) {
    return n = o, t || (t = !0, r = Promise.resolve().then(() => (t = !1, r = void 0, e(...n)))), r;
  };
}
const ge = /* @__PURE__ */ new Set(), Yt = /* @__PURE__ */ new WeakMap();
function Dn() {
  ge.forEach((e) => {
    e(...Yt.get(e));
  }), ge.clear();
}
function Fn(e, ...t) {
  if (typeof e != "function")
    return e;
  Yt.set(e, t), !ge.has(e) && (ge.add(e), ge.size === 1 && Pn(Dn));
}
function Kt(e) {
  return Array.isArray(e) ? e : [e];
}
function K(e, t, n, r) {
  if (!e)
    return B;
  const o = W(() => typeof r == "boolean" ? { capture: r } : r || {});
  let i = B;
  const c = j(
    () => E(e),
    (a) => {
      if (i(), !a)
        return;
      const { disabled: f, ...l } = o.value, u = (v) => {
        E(f) || n(v);
      };
      a.addEventListener(t, u, l), i = () => {
        a.removeEventListener(t, u, l), i = B;
      };
    },
    { immediate: !0, flush: "post" }
  ), s = () => {
    c(), i();
  };
  return Re() && U(s), s;
}
const Ut = "clickoutside", Qe = /* @__PURE__ */ new Set();
let Xe = null;
function _n(e, t) {
  return e instanceof Node ? t === e || t.contains(e) || !!t.__transferElement && (t.__transferElement === e || t.__transferElement.contains(e)) : !1;
}
F && (yt === "click" && document.addEventListener(
  "pointerdown",
  (e) => {
    Xe = e.target;
  },
  !0
), document.addEventListener(
  yt,
  (e) => {
    const t = Xe;
    Xe = null, Qe.forEach((n) => {
      if (t && _n(t, n))
        return;
      const r = e.target, o = e.composedPath && e.composedPath();
      n !== r && (o ? !o.includes(n) : !n.contains(r)) && (!n.__transferElement || n.__transferElement !== r && !n.__transferElement.contains(r)) && Ln(n, { type: Ut });
    });
  },
  !0
));
function go(e, t = C(null)) {
  let n = B;
  const r = j(
    () => E(t),
    (i) => {
      n(), i && (Qe.add(i), n = () => {
        Qe.delete(i), n = B;
      });
    },
    { immediate: !0, flush: "post" }
  ), o = () => {
    r(), n();
  };
  return Re() && U(o), K(t, Ut, e), t;
}
function bo(e = B, t = C(null)) {
  let n;
  return ct(() => {
    Vt(() => {
      const r = Gt(t.value);
      r ? (n = new MutationObserver(() => {
        r.style.display !== "none" && (typeof e == "function" && e(), n == null || n.disconnect(), n = null);
      }), n.observe(r, {
        attributes: !0,
        childList: !0,
        characterData: !0,
        attributeFilter: ["style"]
      })) : typeof e == "function" && e();
    });
  }), at(() => {
    n == null || n.disconnect(), n = null;
  }), t;
}
function Gt(e) {
  if (e) {
    let t = e.parentElement;
    for (; t && t !== document.body; ) {
      if (t.style.display === "none")
        return t;
      t = t.parentElement;
    }
  }
  return null;
}
function xt(e) {
  return (e == null ? void 0 : e.style.display) !== "none" ? !!Gt(e) : !0;
}
function wo(e, t, n) {
  for (const r of t)
    if (e[r])
      return (o) => yn(e, r, o, () => {
        const i = n == null ? void 0 : n(o);
        return i ? Kt(i) : [];
      });
  return n || null;
}
function yo(e) {
  const t = [], n = Array.isArray(e) ? [...e] : [];
  for (; n.length; ) {
    const r = n.shift();
    r !== null && (Array.isArray(r) && n.unshift(...r), !(typeof r != "string" && typeof r != "number" && (!xn(r) || r.type === En)) && (typeof r == "string" || typeof r == "number" ? t.push(Rn(String(r))) : r.type === Sn && Array.isArray(r.children) ? n.unshift(r.children) : t.push(r)));
  }
  return t;
}
function Te(e) {
  const t = E(e);
  return typeof t == "string" ? F ? document.querySelector(t) : null : (t == null ? void 0 : t.$el) ?? t;
}
function xo(e) {
  return new Proxy({}, {
    get(t, n) {
      var r, o, i;
      if (e.component)
        return ((r = e.component.proxy) == null ? void 0 : r[n]) ?? ((o = e.component.exposeProxy) == null ? void 0 : o[n]) ?? ((i = e.component.exposed) == null ? void 0 : i[n]);
    }
  });
}
function Eo(...e) {
  const t = C(!0), n = e[1] || B;
  function r() {
    t.value = !1;
  }
  function o() {
    t.value = !0;
  }
  const i = j(
    e[0],
    (...c) => new Promise((s, a) => {
      t.value && Promise.resolve(n(...c)).then(s).catch(a);
    }),
    e[2]
  );
  return { active: Xt(t), pause: r, resume: o, stop: i };
}
function Ro(e, t) {
  return t.map((n) => On(e, n));
}
const Nn = [
  [
    "requestFullscreen",
    "exitFullscreen",
    "fullscreenElement",
    "fullscreenEnabled",
    "fullscreenchange",
    "fullscreenerror"
  ],
  // New WebKit
  [
    "webkitRequestFullscreen",
    "webkitExitFullscreen",
    "webkitFullscreenElement",
    "webkitFullscreenEnabled",
    "webkitfullscreenchange",
    "webkitfullscreenerror"
  ],
  // Old WebKit
  [
    "webkitRequestFullScreen",
    "webkitCancelFullScreen",
    "webkitCurrentFullScreenElement",
    "webkitCancelFullScreen",
    "webkitfullscreenchange",
    "webkitfullscreenerror"
  ],
  [
    "mozRequestFullScreen",
    "mozCancelFullScreen",
    "mozFullScreenElement",
    "mozFullScreenEnabled",
    "mozfullscreenchange",
    "mozfullscreenerror"
  ],
  [
    "msRequestFullscreen",
    "msExitFullscreen",
    "msFullscreenElement",
    "msFullscreenEnabled",
    "MSFullscreenChange",
    "MSFullscreenError"
  ]
];
let le;
if (F) {
  for (const e of Nn)
    if (e[1] in document) {
      le = e;
      break;
    }
}
const Ze = !!le, Wn = {
  supported: Ze,
  full: W(() => !1),
  enter: B,
  exit: B,
  toggle: B
}, Jt = /* @__PURE__ */ new Set(), et = /* @__PURE__ */ new WeakMap();
if (F && le) {
  const e = le[2], t = le[4];
  document.addEventListener(
    t,
    () => {
      if (Jt.forEach((n) => {
        n.value = !1;
      }), document[e]) {
        const n = et.get(document[e]);
        n && (n.value = !0);
      }
    },
    !1
  );
}
function So(e = C(null)) {
  const t = W({
    get: () => Te(e),
    set: (f) => {
      ae(e) && (e.value = f);
    }
  });
  if (!F || !Ze)
    return { ...Wn, target: t };
  const [n, r, o] = le, i = C(!1);
  j(
    () => Te(e),
    (f, l) => {
      l && et.delete(l), f && et.set(f, i);
    },
    { immediate: !0, flush: "post" }
  ), Jt.add(i), Re() && U(s);
  async function c(f = !1) {
    await s();
    const l = Te(e);
    return l && (f || !document[o]) ? (await l[n](), i.value = !0, document[o] === l) : !1;
  }
  async function s(f = !1) {
    const l = Te(e);
    return f || document[o] && document[o] === l ? (await document[r](), i.value = !1, document[o] !== l) : !1;
  }
  async function a(f = !1) {
    return i.value ? await s(f) : await c(f);
  }
  return {
    supported: Ze,
    target: t,
    full: W(() => i.value),
    enter: c,
    exit: s,
    toggle: a
  };
}
function Oo(e = C(null)) {
  const t = C(!1);
  K(e, "mouseenter", n), K(e, "mouseleave", r);
  function n() {
    t.value = !0;
  }
  function r() {
    t.value = !1;
  }
  return { wrapper: e, isHover: t };
}
const ze = /* @__PURE__ */ new WeakMap(), In = /\s+/g, $n = /(px|%)$/;
function To(e) {
  const t = e.target || C(null);
  if (!F)
    return { target: t, disconnect: B };
  const { handler: n } = e;
  let r = e.root ?? document;
  const o = Kt(e.threshold || 0).join() || "0", i = h(e.rootMargin);
  ze.has(r) || ze.set(r, /* @__PURE__ */ new Map());
  const c = ze.get(r);
  c.has(o) || c.set(o, /* @__PURE__ */ new Map());
  const s = c.get(o);
  s.has(i) || s.set(i, {
    ob: new IntersectionObserver(b, { ...e, rootMargin: i }),
    count: 0,
    handlers: /* @__PURE__ */ new WeakMap()
  });
  let a = s.get(i), { ob: f, handlers: l } = a, u = B;
  const v = j(
    () => E(t),
    (g) => {
      u(), !(!g || !f) && (l.set(g, n), f.observe(g), a && a.count++, u = () => {
        f.unobserve(g), l.delete(g), a && a.count--, u = B;
      });
    },
    { immediate: !0, flush: "post" }
  );
  Re() && U(p);
  function h(g) {
    if (!g || !g.trim()) return "_";
    const m = g.trim().split(In, 4);
    m.length = 4;
    for (let d = 0; d < 4; ++d) {
      const w = m[d];
      m[d] = $n.test(w) ? w : `${kn(w)}px`;
    }
    return m.join(" ");
  }
  function b(g) {
    for (let m = 0, d = g.length; m < d; ++m) {
      const w = g[m], O = l.get(w.target);
      typeof O == "function" && O(w);
    }
  }
  function p() {
    v(), u(), a && (a.count <= 0 && (s.delete(i), s.size || (c.delete(o), c.size || ze.delete(r))), a = void 0, f = void 0, l = void 0, r = void 0);
  }
  return { target: t, disconnect: p };
}
function zo() {
  const e = /* @__PURE__ */ new Set();
  function t(r) {
    return Tn((o, i) => {
      let c = r;
      const s = () => {
        r !== c && (c = r, i());
      };
      return {
        get: () => (o(), r),
        set: (a) => {
          a !== r && (r = a, e.add(s));
        }
      };
    });
  }
  function n() {
    for (const r of e)
      r();
    e.clear();
  }
  return { updateSet: e, manualRef: t, triggerUpdate: n };
}
const Hn = {
  ctrl: "control",
  command: "meta",
  cmd: "meta",
  option: "alt",
  up: "arrowup",
  down: "arrowdown",
  left: "arrowleft",
  right: "arrowright"
}, qn = /[+_-]/, Vn = /[+_-]/g, Xn = ["activeKeys", "resetAll"];
function Co(e = {}) {
  const {
    autoReset: t = !0,
    capture: n = !1,
    passive: r = !0,
    strictTarget: o = !1,
    onKeyDown: i = B,
    onKeyUp: c = B
  } = e, s = e.target || C(null), a = ae(e.disabled) ? e.disabled : C(e.disabled || !1), f = { ...Hn, ...e.aliasMap || {} }, l = Je(/* @__PURE__ */ new Set()), u = /* @__PURE__ */ new Set(), v = Je({ activeKeys: l, resetAll: p });
  function h(m, d) {
    m in v && (v[m] = d);
  }
  function b(m, d) {
    var R, A;
    const w = (R = m.key) == null ? void 0 : R.toLocaleLowerCase(), T = [(A = m.code) == null ? void 0 : A.toLocaleLowerCase(), w].filter(Boolean);
    for (const k of T)
      l[d ? "add" : "delete"](k), h(k, d);
    if (!d && w === "meta") {
      for (const k of u)
        l.delete(k), h(k, !1);
      u.clear();
    } else if (d && typeof m.getModifierState == "function" && m.getModifierState("Meta"))
      for (const k of [...l, ...T])
        u.add(k);
  }
  function p() {
    Object.keys(v).forEach((m) => {
      v[m] = !1;
    }), v.activeKeys = l, v.resetAll = p;
  }
  const g = new Proxy(v, {
    get(m, d, w) {
      if (typeof d != "string" || Xn.includes(d))
        return Reflect.get(m, d, w);
      if (d = d.toLocaleLowerCase(), d in f && (d = f[d]), !(d in v))
        if (qn.test(d)) {
          const O = d.split(Vn).map((T) => T.trim());
          v[d] = W(() => O.every((T) => E(g[T])));
        } else
          v[d] = C(l.has(d));
      return E(Reflect.get(m, d, w));
    }
  });
  return K(
    s,
    "keydown",
    (m) => {
      o && m.target !== E(s) || (b(m, !0), i(m, g));
    },
    { capture: n, passive: r, disabled: a }
  ), K(
    s,
    "keyup",
    (m) => {
      o && m.target !== E(s) || (b(m, !1), c(m, g));
    },
    { capture: n, passive: r, disabled: a }
  ), t && K(s, "blur", p, { capture: n, passive: r, disabled: a }), { target: s, modifier: g };
}
function ko(e) {
  const t = C(!1), n = () => t.value = !0;
  return ct(() => {
    e === "tick" ? Vt(n) : e === "frame" ? requestAnimationFrame(n) : n();
  }), at(() => {
    t.value = !1;
  }), { isMounted: Xt(t) };
}
function je(e) {
  e.cancelable && (e.stopPropagation(), e.preventDefault());
}
function Ao(e) {
  const t = e.target || C(null), n = ae(e.x) ? e.x : C(0), r = ae(e.y) ? e.y : C(0), o = ae(e.lazy) ? e.lazy : C(e.lazy || !1), i = ae(e.disabled) ? e.disabled : C(e.disabled || !1), { capture: c = !0, stopMouse: s = !0, stopTouch: a = !0 } = e, f = C(!1), l = {
    xStart: 0,
    yStart: 0,
    xEnd: 0,
    yEnd: 0,
    clientX: 0,
    clientY: 0,
    deltaX: 0,
    deltaY: 0,
    lazy: !1
  };
  let u = 0, v = 0;
  const h = jt((d) => {
    var w;
    v < u || (m(d), l.lazy || (n.value = l.xEnd, r.value = l.yEnd), (w = e.onMove) == null || w.call(e, l, d));
  });
  function b(d) {
    var w;
    i.value || (Object.assign(l, {
      xStart: n.value,
      yStart: r.value,
      xEnd: n.value,
      yEnd: r.value,
      clientX: d.clientX,
      clientY: d.clientY,
      lazy: o.value
    }), ((w = e.onStart) == null ? void 0 : w.call(e, l, d)) !== !1 && (document.addEventListener("pointermove", p, { capture: c }), document.addEventListener("pointerup", g, { capture: c }), v = u, f.value = !0));
  }
  function p(d) {
    i.value || (je(d), h(d));
  }
  function g(d) {
    var w;
    document.removeEventListener("pointermove", p, { capture: c }), document.removeEventListener("pointerup", g, { capture: c }), !i.value && (m(d), l.lazy && (n.value = l.xEnd, r.value = l.yEnd), f.value = !1, ++u, (w = e.onEnd) == null || w.call(e, l, d));
  }
  function m(d) {
    const { clientX: w, clientY: O } = d, { xStart: T, yStart: R, clientX: A, clientY: k } = l, L = w - A, P = O - k;
    l.deltaX = L, l.deltaY = P, l.xEnd = T + L, l.yEnd = R + P;
  }
  return K(t, "pointerdown", b, { capture: c }), s && K(t, "mousedown", je, { capture: c }), a && K(t, "touchstart", je, { capture: c }), {
    target: t,
    moving: W(() => f.value),
    x: n,
    y: r,
    lazy: o,
    disabled: i
  };
}
const we = /* @__PURE__ */ new Set(), Q = /* @__PURE__ */ new Map();
Q.set("x", 0);
Q.set("y", 0);
function jn(e) {
  const { pageX: t, pageY: n } = e;
  Q.set("x", t), Q.set("y", n), we.forEach((r) => {
    r.x.value = t, r.y.value = n;
  });
}
const Qt = jt(jn);
function Yn(e) {
  !we.size && window && (Q.set("x", 0), Q.set("y", 0), window.addEventListener("pointermove", Qt, { passive: !0 })), we.add(e);
}
function Et(e) {
  we.delete(e), !we.size && window && window.removeEventListener("pointermove", Qt);
}
function Mo(e = {}) {
  const t = C(e.x ?? Q.get("x")), n = C(e.y ?? Q.get("y")), r = { x: t, y: n };
  return Yn(r), e.manualStop || at(() => {
    Et(r);
  }), { ...r, unregister: () => Et(r) };
}
const Kn = ["top", "right", "bottom", "left"], de = Math.min, re = Math.max, De = Math.round, Ce = Math.floor, X = (e) => ({
  x: e,
  y: e
}), Un = {
  left: "right",
  right: "left",
  bottom: "top",
  top: "bottom"
};
function tt(e, t, n) {
  return re(e, de(t, n));
}
function me(e, t) {
  return typeof e == "function" ? e(t) : e;
}
function se(e) {
  return e.split("-")[0];
}
function Se(e) {
  return e.split("-")[1];
}
function Zt(e) {
  return e === "x" ? "y" : "x";
}
function lt(e) {
  return e === "y" ? "height" : "width";
}
function J(e) {
  const t = e[0];
  return t === "t" || t === "b" ? "y" : "x";
}
function ft(e) {
  return Zt(J(e));
}
function Gn(e, t, n) {
  n === void 0 && (n = !1);
  const r = Se(e), o = ft(e), i = lt(o);
  let c = o === "x" ? r === (n ? "end" : "start") ? "right" : "left" : r === "start" ? "bottom" : "top";
  return t.reference[i] > t.floating[i] && (c = Fe(c)), [c, Fe(c)];
}
function Jn(e) {
  const t = Fe(e);
  return [nt(e), t, nt(t)];
}
function nt(e) {
  return e.includes("start") ? e.replace("start", "end") : e.replace("end", "start");
}
const Rt = ["left", "right"], St = ["right", "left"], Qn = ["top", "bottom"], Zn = ["bottom", "top"];
function er(e, t, n) {
  switch (e) {
    case "top":
    case "bottom":
      return n ? t ? St : Rt : t ? Rt : St;
    case "left":
    case "right":
      return t ? Qn : Zn;
    default:
      return [];
  }
}
function tr(e, t, n, r) {
  const o = Se(e);
  let i = er(se(e), n === "start", r);
  return o && (i = i.map((c) => c + "-" + o), t && (i = i.concat(i.map(nt)))), i;
}
function Fe(e) {
  const t = se(e);
  return Un[t] + e.slice(t.length);
}
function nr(e) {
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    ...e
  };
}
function en(e) {
  return typeof e != "number" ? nr(e) : {
    top: e,
    right: e,
    bottom: e,
    left: e
  };
}
function _e(e) {
  const {
    x: t,
    y: n,
    width: r,
    height: o
  } = e;
  return {
    width: r,
    height: o,
    top: n,
    left: t,
    right: t + r,
    bottom: n + o,
    x: t,
    y: n
  };
}
function Ot(e, t, n) {
  let {
    reference: r,
    floating: o
  } = e;
  const i = J(t), c = ft(t), s = lt(c), a = se(t), f = i === "y", l = r.x + r.width / 2 - o.width / 2, u = r.y + r.height / 2 - o.height / 2, v = r[s] / 2 - o[s] / 2;
  let h;
  switch (a) {
    case "top":
      h = {
        x: l,
        y: r.y - o.height
      };
      break;
    case "bottom":
      h = {
        x: l,
        y: r.y + r.height
      };
      break;
    case "right":
      h = {
        x: r.x + r.width,
        y: u
      };
      break;
    case "left":
      h = {
        x: r.x - o.width,
        y: u
      };
      break;
    default:
      h = {
        x: r.x,
        y: r.y
      };
  }
  switch (Se(t)) {
    case "start":
      h[c] -= v * (n && f ? -1 : 1);
      break;
    case "end":
      h[c] += v * (n && f ? -1 : 1);
      break;
  }
  return h;
}
async function rr(e, t) {
  var n;
  t === void 0 && (t = {});
  const {
    x: r,
    y: o,
    platform: i,
    rects: c,
    elements: s,
    strategy: a
  } = e, {
    boundary: f = "clippingAncestors",
    rootBoundary: l = "viewport",
    elementContext: u = "floating",
    altBoundary: v = !1,
    padding: h = 0
  } = me(t, e), b = en(h), g = s[v ? u === "floating" ? "reference" : "floating" : u], m = _e(await i.getClippingRect({
    element: (n = await (i.isElement == null ? void 0 : i.isElement(g))) == null || n ? g : g.contextElement || await (i.getDocumentElement == null ? void 0 : i.getDocumentElement(s.floating)),
    boundary: f,
    rootBoundary: l,
    strategy: a
  })), d = u === "floating" ? {
    x: r,
    y: o,
    width: c.floating.width,
    height: c.floating.height
  } : c.reference, w = await (i.getOffsetParent == null ? void 0 : i.getOffsetParent(s.floating)), O = await (i.isElement == null ? void 0 : i.isElement(w)) ? await (i.getScale == null ? void 0 : i.getScale(w)) || {
    x: 1,
    y: 1
  } : {
    x: 1,
    y: 1
  }, T = _e(i.convertOffsetParentRelativeRectToViewportRelativeRect ? await i.convertOffsetParentRelativeRectToViewportRelativeRect({
    elements: s,
    rect: d,
    offsetParent: w,
    strategy: a
  }) : d);
  return {
    top: (m.top - T.top + b.top) / O.y,
    bottom: (T.bottom - m.bottom + b.bottom) / O.y,
    left: (m.left - T.left + b.left) / O.x,
    right: (T.right - m.right + b.right) / O.x
  };
}
const or = 50, ir = async (e, t, n) => {
  const {
    placement: r = "bottom",
    strategy: o = "absolute",
    middleware: i = [],
    platform: c
  } = n, s = c.detectOverflow ? c : {
    ...c,
    detectOverflow: rr
  }, a = await (c.isRTL == null ? void 0 : c.isRTL(t));
  let f = await c.getElementRects({
    reference: e,
    floating: t,
    strategy: o
  }), {
    x: l,
    y: u
  } = Ot(f, r, a), v = r, h = 0;
  const b = {};
  for (let p = 0; p < i.length; p++) {
    const g = i[p];
    if (!g)
      continue;
    const {
      name: m,
      fn: d
    } = g, {
      x: w,
      y: O,
      data: T,
      reset: R
    } = await d({
      x: l,
      y: u,
      initialPlacement: r,
      placement: v,
      strategy: o,
      middlewareData: b,
      rects: f,
      platform: s,
      elements: {
        reference: e,
        floating: t
      }
    });
    l = w ?? l, u = O ?? u, b[m] = {
      ...b[m],
      ...T
    }, R && h < or && (h++, typeof R == "object" && (R.placement && (v = R.placement), R.rects && (f = R.rects === !0 ? await c.getElementRects({
      reference: e,
      floating: t,
      strategy: o
    }) : R.rects), {
      x: l,
      y: u
    } = Ot(f, v, a)), p = -1);
  }
  return {
    x: l,
    y: u,
    placement: v,
    strategy: o,
    middlewareData: b
  };
}, sr = (e) => ({
  name: "arrow",
  options: e,
  async fn(t) {
    const {
      x: n,
      y: r,
      placement: o,
      rects: i,
      platform: c,
      elements: s,
      middlewareData: a
    } = t, {
      element: f,
      padding: l = 0
    } = me(e, t) || {};
    if (f == null)
      return {};
    const u = en(l), v = {
      x: n,
      y: r
    }, h = ft(o), b = lt(h), p = await c.getDimensions(f), g = h === "y", m = g ? "top" : "left", d = g ? "bottom" : "right", w = g ? "clientHeight" : "clientWidth", O = i.reference[b] + i.reference[h] - v[h] - i.floating[b], T = v[h] - i.reference[h], R = await (c.getOffsetParent == null ? void 0 : c.getOffsetParent(f));
    let A = R ? R[w] : 0;
    (!A || !await (c.isElement == null ? void 0 : c.isElement(R))) && (A = s.floating[w] || i.floating[b]);
    const k = O / 2 - T / 2, L = A / 2 - p[b] / 2 - 1, P = de(u[m], L), $ = de(u[d], L), N = P, y = A - p[b] - $, x = A / 2 - p[b] / 2 + k, S = tt(N, x, y), M = !a.arrow && Se(o) != null && x !== S && i.reference[b] / 2 - (x < N ? P : $) - p[b] / 2 < 0, z = M ? x < N ? x - N : x - y : 0;
    return {
      [h]: v[h] + z,
      data: {
        [h]: S,
        centerOffset: x - S - z,
        ...M && {
          alignmentOffset: z
        }
      },
      reset: M
    };
  }
}), cr = function(e) {
  return e === void 0 && (e = {}), {
    name: "flip",
    options: e,
    async fn(t) {
      var n, r;
      const {
        placement: o,
        middlewareData: i,
        rects: c,
        initialPlacement: s,
        platform: a,
        elements: f
      } = t, {
        mainAxis: l = !0,
        crossAxis: u = !0,
        fallbackPlacements: v,
        fallbackStrategy: h = "bestFit",
        fallbackAxisSideDirection: b = "none",
        flipAlignment: p = !0,
        ...g
      } = me(e, t);
      if ((n = i.arrow) != null && n.alignmentOffset)
        return {};
      const m = se(o), d = J(s), w = se(s) === s, O = await (a.isRTL == null ? void 0 : a.isRTL(f.floating)), T = v || (w || !p ? [Fe(s)] : Jn(s)), R = b !== "none";
      !v && R && T.push(...tr(s, p, b, O));
      const A = [s, ...T], k = await a.detectOverflow(t, g), L = [];
      let P = ((r = i.flip) == null ? void 0 : r.overflows) || [];
      if (l && L.push(k[m]), u) {
        const x = Gn(o, c, O);
        L.push(k[x[0]], k[x[1]]);
      }
      if (P = [...P, {
        placement: o,
        overflows: L
      }], !L.every((x) => x <= 0)) {
        var $, N;
        const x = ((($ = i.flip) == null ? void 0 : $.index) || 0) + 1, S = A[x];
        if (S && (!(u === "alignment" ? d !== J(S) : !1) || // We leave the current main axis only if every placement on that axis
        // overflows the main axis.
        P.every((D) => J(D.placement) === d ? D.overflows[0] > 0 : !0)))
          return {
            data: {
              index: x,
              overflows: P
            },
            reset: {
              placement: S
            }
          };
        let M = (N = P.filter((z) => z.overflows[0] <= 0).sort((z, D) => z.overflows[1] - D.overflows[1])[0]) == null ? void 0 : N.placement;
        if (!M)
          switch (h) {
            case "bestFit": {
              var y;
              const z = (y = P.filter((D) => {
                if (R) {
                  const _ = J(D.placement);
                  return _ === d || // Create a bias to the `y` side axis due to horizontal
                  // reading directions favoring greater width.
                  _ === "y";
                }
                return !0;
              }).map((D) => [D.placement, D.overflows.filter((_) => _ > 0).reduce((_, ee) => _ + ee, 0)]).sort((D, _) => D[1] - _[1])[0]) == null ? void 0 : y[0];
              z && (M = z);
              break;
            }
            case "initialPlacement":
              M = s;
              break;
          }
        if (o !== M)
          return {
            reset: {
              placement: M
            }
          };
      }
      return {};
    }
  };
};
function Tt(e, t) {
  return {
    top: e.top - t.height,
    right: e.right - t.width,
    bottom: e.bottom - t.height,
    left: e.left - t.width
  };
}
function zt(e) {
  return Kn.some((t) => e[t] >= 0);
}
const ar = function(e) {
  return e === void 0 && (e = {}), {
    name: "hide",
    options: e,
    async fn(t) {
      const {
        rects: n,
        platform: r
      } = t, {
        strategy: o = "referenceHidden",
        ...i
      } = me(e, t);
      switch (o) {
        case "referenceHidden": {
          const c = await r.detectOverflow(t, {
            ...i,
            elementContext: "reference"
          }), s = Tt(c, n.reference);
          return {
            data: {
              referenceHiddenOffsets: s,
              referenceHidden: zt(s)
            }
          };
        }
        case "escaped": {
          const c = await r.detectOverflow(t, {
            ...i,
            altBoundary: !0
          }), s = Tt(c, n.floating);
          return {
            data: {
              escapedOffsets: s,
              escaped: zt(s)
            }
          };
        }
        default:
          return {};
      }
    }
  };
}, lr = /* @__PURE__ */ new Set(["left", "top"]);
async function fr(e, t) {
  const {
    placement: n,
    platform: r,
    elements: o
  } = e, i = await (r.isRTL == null ? void 0 : r.isRTL(o.floating)), c = se(n), s = Se(n), a = J(n) === "y", f = lr.has(c) ? -1 : 1, l = i && a ? -1 : 1, u = me(t, e);
  let {
    mainAxis: v,
    crossAxis: h,
    alignmentAxis: b
  } = typeof u == "number" ? {
    mainAxis: u,
    crossAxis: 0,
    alignmentAxis: null
  } : {
    mainAxis: u.mainAxis || 0,
    crossAxis: u.crossAxis || 0,
    alignmentAxis: u.alignmentAxis
  };
  return s && typeof b == "number" && (h = s === "end" ? b * -1 : b), a ? {
    x: h * l,
    y: v * f
  } : {
    x: v * f,
    y: h * l
  };
}
const ur = function(e) {
  return e === void 0 && (e = 0), {
    name: "offset",
    options: e,
    async fn(t) {
      var n, r;
      const {
        x: o,
        y: i,
        placement: c,
        middlewareData: s
      } = t, a = await fr(t, e);
      return c === ((n = s.offset) == null ? void 0 : n.placement) && (r = s.arrow) != null && r.alignmentOffset ? {} : {
        x: o + a.x,
        y: i + a.y,
        data: {
          ...a,
          placement: c
        }
      };
    }
  };
}, dr = function(e) {
  return e === void 0 && (e = {}), {
    name: "shift",
    options: e,
    async fn(t) {
      const {
        x: n,
        y: r,
        placement: o,
        platform: i
      } = t, {
        mainAxis: c = !0,
        crossAxis: s = !1,
        limiter: a = {
          fn: (m) => {
            let {
              x: d,
              y: w
            } = m;
            return {
              x: d,
              y: w
            };
          }
        },
        ...f
      } = me(e, t), l = {
        x: n,
        y: r
      }, u = await i.detectOverflow(t, f), v = J(se(o)), h = Zt(v);
      let b = l[h], p = l[v];
      if (c) {
        const m = h === "y" ? "top" : "left", d = h === "y" ? "bottom" : "right", w = b + u[m], O = b - u[d];
        b = tt(w, b, O);
      }
      if (s) {
        const m = v === "y" ? "top" : "left", d = v === "y" ? "bottom" : "right", w = p + u[m], O = p - u[d];
        p = tt(w, p, O);
      }
      const g = a.fn({
        ...t,
        [h]: b,
        [v]: p
      });
      return {
        ...g,
        data: {
          x: g.x - n,
          y: g.y - r,
          enabled: {
            [h]: c,
            [v]: s
          }
        }
      };
    }
  };
};
function Ie() {
  return typeof window < "u";
}
function ve(e) {
  return tn(e) ? (e.nodeName || "").toLowerCase() : "#document";
}
function I(e) {
  var t;
  return (e == null || (t = e.ownerDocument) == null ? void 0 : t.defaultView) || window;
}
function Y(e) {
  var t;
  return (t = (tn(e) ? e.ownerDocument : e.document) || window.document) == null ? void 0 : t.documentElement;
}
function tn(e) {
  return Ie() ? e instanceof Node || e instanceof I(e).Node : !1;
}
function H(e) {
  return Ie() ? e instanceof Element || e instanceof I(e).Element : !1;
}
function G(e) {
  return Ie() ? e instanceof HTMLElement || e instanceof I(e).HTMLElement : !1;
}
function Ct(e) {
  return !Ie() || typeof ShadowRoot > "u" ? !1 : e instanceof ShadowRoot || e instanceof I(e).ShadowRoot;
}
function Oe(e) {
  const {
    overflow: t,
    overflowX: n,
    overflowY: r,
    display: o
  } = q(e);
  return /auto|scroll|overlay|hidden|clip/.test(t + r + n) && o !== "inline" && o !== "contents";
}
function hr(e) {
  return /^(table|td|th)$/.test(ve(e));
}
function $e(e) {
  try {
    if (e.matches(":popover-open"))
      return !0;
  } catch {
  }
  try {
    return e.matches(":modal");
  } catch {
    return !1;
  }
}
const mr = /transform|translate|scale|rotate|perspective|filter/, vr = /paint|layout|strict|content/, ne = (e) => !!e && e !== "none";
let Ye;
function ut(e) {
  const t = H(e) ? q(e) : e;
  return ne(t.transform) || ne(t.translate) || ne(t.scale) || ne(t.rotate) || ne(t.perspective) || !dt() && (ne(t.backdropFilter) || ne(t.filter)) || mr.test(t.willChange || "") || vr.test(t.contain || "");
}
function pr(e) {
  let t = Z(e);
  for (; G(t) && !he(t); ) {
    if (ut(t))
      return t;
    if ($e(t))
      return null;
    t = Z(t);
  }
  return null;
}
function dt() {
  return Ye == null && (Ye = typeof CSS < "u" && CSS.supports && CSS.supports("-webkit-backdrop-filter", "none")), Ye;
}
function he(e) {
  return /^(html|body|#document)$/.test(ve(e));
}
function q(e) {
  return I(e).getComputedStyle(e);
}
function He(e) {
  return H(e) ? {
    scrollLeft: e.scrollLeft,
    scrollTop: e.scrollTop
  } : {
    scrollLeft: e.scrollX,
    scrollTop: e.scrollY
  };
}
function Z(e) {
  if (ve(e) === "html")
    return e;
  const t = (
    // Step into the shadow DOM of the parent of a slotted node.
    e.assignedSlot || // DOM Element detected.
    e.parentNode || // ShadowRoot detected.
    Ct(e) && e.host || // Fallback.
    Y(e)
  );
  return Ct(t) ? t.host : t;
}
function nn(e) {
  const t = Z(e);
  return he(t) ? e.ownerDocument ? e.ownerDocument.body : e.body : G(t) && Oe(t) ? t : nn(t);
}
function ye(e, t, n) {
  var r;
  t === void 0 && (t = []), n === void 0 && (n = !0);
  const o = nn(e), i = o === ((r = e.ownerDocument) == null ? void 0 : r.body), c = I(o);
  if (i) {
    const s = rt(c);
    return t.concat(c, c.visualViewport || [], Oe(o) ? o : [], s && n ? ye(s) : []);
  } else
    return t.concat(o, ye(o, [], n));
}
function rt(e) {
  return e.parent && Object.getPrototypeOf(e.parent) ? e.frameElement : null;
}
function rn(e) {
  const t = q(e);
  let n = parseFloat(t.width) || 0, r = parseFloat(t.height) || 0;
  const o = G(e), i = o ? e.offsetWidth : n, c = o ? e.offsetHeight : r, s = De(n) !== i || De(r) !== c;
  return s && (n = i, r = c), {
    width: n,
    height: r,
    $: s
  };
}
function ht(e) {
  return H(e) ? e : e.contextElement;
}
function fe(e) {
  const t = ht(e);
  if (!G(t))
    return X(1);
  const n = t.getBoundingClientRect(), {
    width: r,
    height: o,
    $: i
  } = rn(t);
  let c = (i ? De(n.width) : n.width) / r, s = (i ? De(n.height) : n.height) / o;
  return (!c || !Number.isFinite(c)) && (c = 1), (!s || !Number.isFinite(s)) && (s = 1), {
    x: c,
    y: s
  };
}
const gr = /* @__PURE__ */ X(0);
function on(e) {
  const t = I(e);
  return !dt() || !t.visualViewport ? gr : {
    x: t.visualViewport.offsetLeft,
    y: t.visualViewport.offsetTop
  };
}
function br(e, t, n) {
  return t === void 0 && (t = !1), !n || t && n !== I(e) ? !1 : t;
}
function ce(e, t, n, r) {
  t === void 0 && (t = !1), n === void 0 && (n = !1);
  const o = e.getBoundingClientRect(), i = ht(e);
  let c = X(1);
  t && (r ? H(r) && (c = fe(r)) : c = fe(e));
  const s = br(i, n, r) ? on(i) : X(0);
  let a = (o.left + s.x) / c.x, f = (o.top + s.y) / c.y, l = o.width / c.x, u = o.height / c.y;
  if (i) {
    const v = I(i), h = r && H(r) ? I(r) : r;
    let b = v, p = rt(b);
    for (; p && r && h !== b; ) {
      const g = fe(p), m = p.getBoundingClientRect(), d = q(p), w = m.left + (p.clientLeft + parseFloat(d.paddingLeft)) * g.x, O = m.top + (p.clientTop + parseFloat(d.paddingTop)) * g.y;
      a *= g.x, f *= g.y, l *= g.x, u *= g.y, a += w, f += O, b = I(p), p = rt(b);
    }
  }
  return _e({
    width: l,
    height: u,
    x: a,
    y: f
  });
}
function qe(e, t) {
  const n = He(e).scrollLeft;
  return t ? t.left + n : ce(Y(e)).left + n;
}
function sn(e, t) {
  const n = e.getBoundingClientRect(), r = n.left + t.scrollLeft - qe(e, n), o = n.top + t.scrollTop;
  return {
    x: r,
    y: o
  };
}
function wr(e) {
  let {
    elements: t,
    rect: n,
    offsetParent: r,
    strategy: o
  } = e;
  const i = o === "fixed", c = Y(r), s = t ? $e(t.floating) : !1;
  if (r === c || s && i)
    return n;
  let a = {
    scrollLeft: 0,
    scrollTop: 0
  }, f = X(1);
  const l = X(0), u = G(r);
  if ((u || !u && !i) && ((ve(r) !== "body" || Oe(c)) && (a = He(r)), u)) {
    const h = ce(r);
    f = fe(r), l.x = h.x + r.clientLeft, l.y = h.y + r.clientTop;
  }
  const v = c && !u && !i ? sn(c, a) : X(0);
  return {
    width: n.width * f.x,
    height: n.height * f.y,
    x: n.x * f.x - a.scrollLeft * f.x + l.x + v.x,
    y: n.y * f.y - a.scrollTop * f.y + l.y + v.y
  };
}
function yr(e) {
  return Array.from(e.getClientRects());
}
function xr(e) {
  const t = Y(e), n = He(e), r = e.ownerDocument.body, o = re(t.scrollWidth, t.clientWidth, r.scrollWidth, r.clientWidth), i = re(t.scrollHeight, t.clientHeight, r.scrollHeight, r.clientHeight);
  let c = -n.scrollLeft + qe(e);
  const s = -n.scrollTop;
  return q(r).direction === "rtl" && (c += re(t.clientWidth, r.clientWidth) - o), {
    width: o,
    height: i,
    x: c,
    y: s
  };
}
const kt = 25;
function Er(e, t) {
  const n = I(e), r = Y(e), o = n.visualViewport;
  let i = r.clientWidth, c = r.clientHeight, s = 0, a = 0;
  if (o) {
    i = o.width, c = o.height;
    const l = dt();
    (!l || l && t === "fixed") && (s = o.offsetLeft, a = o.offsetTop);
  }
  const f = qe(r);
  if (f <= 0) {
    const l = r.ownerDocument, u = l.body, v = getComputedStyle(u), h = l.compatMode === "CSS1Compat" && parseFloat(v.marginLeft) + parseFloat(v.marginRight) || 0, b = Math.abs(r.clientWidth - u.clientWidth - h);
    b <= kt && (i -= b);
  } else f <= kt && (i += f);
  return {
    width: i,
    height: c,
    x: s,
    y: a
  };
}
function Rr(e, t) {
  const n = ce(e, !0, t === "fixed"), r = n.top + e.clientTop, o = n.left + e.clientLeft, i = G(e) ? fe(e) : X(1), c = e.clientWidth * i.x, s = e.clientHeight * i.y, a = o * i.x, f = r * i.y;
  return {
    width: c,
    height: s,
    x: a,
    y: f
  };
}
function At(e, t, n) {
  let r;
  if (t === "viewport")
    r = Er(e, n);
  else if (t === "document")
    r = xr(Y(e));
  else if (H(t))
    r = Rr(t, n);
  else {
    const o = on(e);
    r = {
      x: t.x - o.x,
      y: t.y - o.y,
      width: t.width,
      height: t.height
    };
  }
  return _e(r);
}
function cn(e, t) {
  const n = Z(e);
  return n === t || !H(n) || he(n) ? !1 : q(n).position === "fixed" || cn(n, t);
}
function Sr(e, t) {
  const n = t.get(e);
  if (n)
    return n;
  let r = ye(e, [], !1).filter((s) => H(s) && ve(s) !== "body"), o = null;
  const i = q(e).position === "fixed";
  let c = i ? Z(e) : e;
  for (; H(c) && !he(c); ) {
    const s = q(c), a = ut(c);
    !a && s.position === "fixed" && (o = null), (i ? !a && !o : !a && s.position === "static" && !!o && (o.position === "absolute" || o.position === "fixed") || Oe(c) && !a && cn(e, c)) ? r = r.filter((l) => l !== c) : o = s, c = Z(c);
  }
  return t.set(e, r), r;
}
function Or(e) {
  let {
    element: t,
    boundary: n,
    rootBoundary: r,
    strategy: o
  } = e;
  const c = [...n === "clippingAncestors" ? $e(t) ? [] : Sr(t, this._c) : [].concat(n), r], s = At(t, c[0], o);
  let a = s.top, f = s.right, l = s.bottom, u = s.left;
  for (let v = 1; v < c.length; v++) {
    const h = At(t, c[v], o);
    a = re(h.top, a), f = de(h.right, f), l = de(h.bottom, l), u = re(h.left, u);
  }
  return {
    width: f - u,
    height: l - a,
    x: u,
    y: a
  };
}
function Tr(e) {
  const {
    width: t,
    height: n
  } = rn(e);
  return {
    width: t,
    height: n
  };
}
function zr(e, t, n) {
  const r = G(t), o = Y(t), i = n === "fixed", c = ce(e, !0, i, t);
  let s = {
    scrollLeft: 0,
    scrollTop: 0
  };
  const a = X(0);
  function f() {
    a.x = qe(o);
  }
  if (r || !r && !i)
    if ((ve(t) !== "body" || Oe(o)) && (s = He(t)), r) {
      const h = ce(t, !0, i, t);
      a.x = h.x + t.clientLeft, a.y = h.y + t.clientTop;
    } else o && f();
  i && !r && o && f();
  const l = o && !r && !i ? sn(o, s) : X(0), u = c.left + s.scrollLeft - a.x - l.x, v = c.top + s.scrollTop - a.y - l.y;
  return {
    x: u,
    y: v,
    width: c.width,
    height: c.height
  };
}
function Ke(e) {
  return q(e).position === "static";
}
function Mt(e, t) {
  if (!G(e) || q(e).position === "fixed")
    return null;
  if (t)
    return t(e);
  let n = e.offsetParent;
  return Y(e) === n && (n = n.ownerDocument.body), n;
}
function an(e, t) {
  const n = I(e);
  if ($e(e))
    return n;
  if (!G(e)) {
    let o = Z(e);
    for (; o && !he(o); ) {
      if (H(o) && !Ke(o))
        return o;
      o = Z(o);
    }
    return n;
  }
  let r = Mt(e, t);
  for (; r && hr(r) && Ke(r); )
    r = Mt(r, t);
  return r && he(r) && Ke(r) && !ut(r) ? n : r || pr(e) || n;
}
const Cr = async function(e) {
  const t = this.getOffsetParent || an, n = this.getDimensions, r = await n(e.floating);
  return {
    reference: zr(e.reference, await t(e.floating), e.strategy),
    floating: {
      x: 0,
      y: 0,
      width: r.width,
      height: r.height
    }
  };
};
function kr(e) {
  return q(e).direction === "rtl";
}
const ln = {
  convertOffsetParentRelativeRectToViewportRelativeRect: wr,
  getDocumentElement: Y,
  getClippingRect: Or,
  getOffsetParent: an,
  getElementRects: Cr,
  getClientRects: yr,
  getDimensions: Tr,
  getScale: fe,
  isElement: H,
  isRTL: kr
};
function fn(e, t) {
  return e.x === t.x && e.y === t.y && e.width === t.width && e.height === t.height;
}
function Ar(e, t) {
  let n = null, r;
  const o = Y(e);
  function i() {
    var s;
    clearTimeout(r), (s = n) == null || s.disconnect(), n = null;
  }
  function c(s, a) {
    s === void 0 && (s = !1), a === void 0 && (a = 1), i();
    const f = e.getBoundingClientRect(), {
      left: l,
      top: u,
      width: v,
      height: h
    } = f;
    if (s || t(), !v || !h)
      return;
    const b = Ce(u), p = Ce(o.clientWidth - (l + v)), g = Ce(o.clientHeight - (u + h)), m = Ce(l), w = {
      rootMargin: -b + "px " + -p + "px " + -g + "px " + -m + "px",
      threshold: re(0, de(1, a)) || 1
    };
    let O = !0;
    function T(R) {
      const A = R[0].intersectionRatio;
      if (A !== a) {
        if (!O)
          return c();
        A ? c(!1, A) : r = setTimeout(() => {
          c(!1, 1e-7);
        }, 1e3);
      }
      A === 1 && !fn(f, e.getBoundingClientRect()) && c(), O = !1;
    }
    try {
      n = new IntersectionObserver(T, {
        ...w,
        // Handle <iframe>s
        root: o.ownerDocument
      });
    } catch {
      n = new IntersectionObserver(T, w);
    }
    n.observe(e);
  }
  return c(!0), i;
}
function Mr(e, t, n, r) {
  r === void 0 && (r = {});
  const {
    ancestorScroll: o = !0,
    ancestorResize: i = !0,
    elementResize: c = typeof ResizeObserver == "function",
    layoutShift: s = typeof IntersectionObserver == "function",
    animationFrame: a = !1
  } = r, f = ht(e), l = o || i ? [...f ? ye(f) : [], ...t ? ye(t) : []] : [];
  l.forEach((m) => {
    o && m.addEventListener("scroll", n, {
      passive: !0
    }), i && m.addEventListener("resize", n);
  });
  const u = f && s ? Ar(f, n) : null;
  let v = -1, h = null;
  c && (h = new ResizeObserver((m) => {
    let [d] = m;
    d && d.target === f && h && t && (h.unobserve(t), cancelAnimationFrame(v), v = requestAnimationFrame(() => {
      var w;
      (w = h) == null || w.observe(t);
    })), n();
  }), f && !a && h.observe(f), t && h.observe(t));
  let b, p = a ? ce(e) : null;
  a && g();
  function g() {
    const m = ce(e);
    p && !fn(p, m) && n(), p = m, b = requestAnimationFrame(g);
  }
  return n(), () => {
    var m;
    l.forEach((d) => {
      o && d.removeEventListener("scroll", n), i && d.removeEventListener("resize", n);
    }), u == null || u(), (m = h) == null || m.disconnect(), h = null, a && cancelAnimationFrame(b);
  };
}
const Lr = ur, Pr = dr, Br = cr, Dr = ar, Fr = sr, _r = (e, t, n) => {
  const r = /* @__PURE__ */ new Map(), o = {
    platform: ln,
    ...n
  }, i = {
    ...o.platform,
    _c: r
  };
  return ir(e, t, {
    ...o,
    platform: i
  });
};
var oe = [], Nr = function() {
  return oe.some(function(e) {
    return e.activeTargets.length > 0;
  });
}, Wr = function() {
  return oe.some(function(e) {
    return e.skippedTargets.length > 0;
  });
}, Lt = "ResizeObserver loop completed with undelivered notifications.", Ir = function() {
  var e;
  typeof ErrorEvent == "function" ? e = new ErrorEvent("error", {
    message: Lt
  }) : (e = document.createEvent("Event"), e.initEvent("error", !1, !1), e.message = Lt), window.dispatchEvent(e);
}, xe;
(function(e) {
  e.BORDER_BOX = "border-box", e.CONTENT_BOX = "content-box", e.DEVICE_PIXEL_CONTENT_BOX = "device-pixel-content-box";
})(xe || (xe = {}));
var ie = function(e) {
  return Object.freeze(e);
}, $r = /* @__PURE__ */ (function() {
  function e(t, n) {
    this.inlineSize = t, this.blockSize = n, ie(this);
  }
  return e;
})(), un = (function() {
  function e(t, n, r, o) {
    return this.x = t, this.y = n, this.width = r, this.height = o, this.top = this.y, this.left = this.x, this.bottom = this.top + this.height, this.right = this.left + this.width, ie(this);
  }
  return e.prototype.toJSON = function() {
    var t = this, n = t.x, r = t.y, o = t.top, i = t.right, c = t.bottom, s = t.left, a = t.width, f = t.height;
    return { x: n, y: r, top: o, right: i, bottom: c, left: s, width: a, height: f };
  }, e.fromRect = function(t) {
    return new e(t.x, t.y, t.width, t.height);
  }, e;
})(), mt = function(e) {
  return e instanceof SVGElement && "getBBox" in e;
}, dn = function(e) {
  if (mt(e)) {
    var t = e.getBBox(), n = t.width, r = t.height;
    return !n && !r;
  }
  var o = e, i = o.offsetWidth, c = o.offsetHeight;
  return !(i || c || e.getClientRects().length);
}, Pt = function(e) {
  var t;
  if (e instanceof Element)
    return !0;
  var n = (t = e == null ? void 0 : e.ownerDocument) === null || t === void 0 ? void 0 : t.defaultView;
  return !!(n && e instanceof n.Element);
}, Hr = function(e) {
  switch (e.tagName) {
    case "INPUT":
      if (e.type !== "image")
        break;
    case "VIDEO":
    case "AUDIO":
    case "EMBED":
    case "OBJECT":
    case "CANVAS":
    case "IFRAME":
    case "IMG":
      return !0;
  }
  return !1;
}, be = typeof window < "u" ? window : {}, ke = /* @__PURE__ */ new WeakMap(), Bt = /auto|scroll/, qr = /^tb|vertical/, Vr = /msie|trident/i.test(be.navigator && be.navigator.userAgent), V = function(e) {
  return parseFloat(e || "0");
}, ue = function(e, t, n) {
  return e === void 0 && (e = 0), t === void 0 && (t = 0), n === void 0 && (n = !1), new $r((n ? t : e) || 0, (n ? e : t) || 0);
}, Dt = ie({
  devicePixelContentBoxSize: ue(),
  borderBoxSize: ue(),
  contentBoxSize: ue(),
  contentRect: new un(0, 0, 0, 0)
}), hn = function(e, t) {
  if (t === void 0 && (t = !1), ke.has(e) && !t)
    return ke.get(e);
  if (dn(e))
    return ke.set(e, Dt), Dt;
  var n = getComputedStyle(e), r = mt(e) && e.ownerSVGElement && e.getBBox(), o = !Vr && n.boxSizing === "border-box", i = qr.test(n.writingMode || ""), c = !r && Bt.test(n.overflowY || ""), s = !r && Bt.test(n.overflowX || ""), a = r ? 0 : V(n.paddingTop), f = r ? 0 : V(n.paddingRight), l = r ? 0 : V(n.paddingBottom), u = r ? 0 : V(n.paddingLeft), v = r ? 0 : V(n.borderTopWidth), h = r ? 0 : V(n.borderRightWidth), b = r ? 0 : V(n.borderBottomWidth), p = r ? 0 : V(n.borderLeftWidth), g = u + f, m = a + l, d = p + h, w = v + b, O = s ? e.offsetHeight - w - e.clientHeight : 0, T = c ? e.offsetWidth - d - e.clientWidth : 0, R = o ? g + d : 0, A = o ? m + w : 0, k = r ? r.width : V(n.width) - R - T, L = r ? r.height : V(n.height) - A - O, P = k + g + T + d, $ = L + m + O + w, N = ie({
    devicePixelContentBoxSize: ue(Math.round(k * devicePixelRatio), Math.round(L * devicePixelRatio), i),
    borderBoxSize: ue(P, $, i),
    contentBoxSize: ue(k, L, i),
    contentRect: new un(u, a, k, L)
  });
  return ke.set(e, N), N;
}, mn = function(e, t, n) {
  var r = hn(e, n), o = r.borderBoxSize, i = r.contentBoxSize, c = r.devicePixelContentBoxSize;
  switch (t) {
    case xe.DEVICE_PIXEL_CONTENT_BOX:
      return c;
    case xe.BORDER_BOX:
      return o;
    default:
      return i;
  }
}, Xr = /* @__PURE__ */ (function() {
  function e(t) {
    var n = hn(t);
    this.target = t, this.contentRect = n.contentRect, this.borderBoxSize = ie([n.borderBoxSize]), this.contentBoxSize = ie([n.contentBoxSize]), this.devicePixelContentBoxSize = ie([n.devicePixelContentBoxSize]);
  }
  return e;
})(), vn = function(e) {
  if (dn(e))
    return 1 / 0;
  for (var t = 0, n = e.parentNode; n; )
    t += 1, n = n.parentNode;
  return t;
}, jr = function() {
  var e = 1 / 0, t = [];
  oe.forEach(function(c) {
    if (c.activeTargets.length !== 0) {
      var s = [];
      c.activeTargets.forEach(function(f) {
        var l = new Xr(f.target), u = vn(f.target);
        s.push(l), f.lastReportedSize = mn(f.target, f.observedBox), u < e && (e = u);
      }), t.push(function() {
        c.callback.call(c.observer, s, c.observer);
      }), c.activeTargets.splice(0, c.activeTargets.length);
    }
  });
  for (var n = 0, r = t; n < r.length; n++) {
    var o = r[n];
    o();
  }
  return e;
}, Ft = function(e) {
  oe.forEach(function(n) {
    n.activeTargets.splice(0, n.activeTargets.length), n.skippedTargets.splice(0, n.skippedTargets.length), n.observationTargets.forEach(function(o) {
      o.isActive() && (vn(o.target) > e ? n.activeTargets.push(o) : n.skippedTargets.push(o));
    });
  });
}, Yr = function() {
  var e = 0;
  for (Ft(e); Nr(); )
    e = jr(), Ft(e);
  return Wr() && Ir(), e > 0;
}, Ue, pn = [], Kr = function() {
  return pn.splice(0).forEach(function(e) {
    return e();
  });
}, Ur = function(e) {
  if (!Ue) {
    var t = 0, n = document.createTextNode(""), r = { characterData: !0 };
    new MutationObserver(function() {
      return Kr();
    }).observe(n, r), Ue = function() {
      n.textContent = "".concat(t ? t-- : t++);
    };
  }
  pn.push(e), Ue();
}, Gr = function(e) {
  Ur(function() {
    requestAnimationFrame(e);
  });
}, Le = 0, Jr = function() {
  return !!Le;
}, Qr = 250, Zr = { attributes: !0, characterData: !0, childList: !0, subtree: !0 }, _t = [
  "resize",
  "load",
  "transitionend",
  "animationend",
  "animationstart",
  "animationiteration",
  "keyup",
  "keydown",
  "mouseup",
  "mousedown",
  "mouseover",
  "mouseout",
  "blur",
  "focus"
], Nt = function(e) {
  return e === void 0 && (e = 0), Date.now() + e;
}, Ge = !1, eo = (function() {
  function e() {
    var t = this;
    this.stopped = !0, this.listener = function() {
      return t.schedule();
    };
  }
  return e.prototype.run = function(t) {
    var n = this;
    if (t === void 0 && (t = Qr), !Ge) {
      Ge = !0;
      var r = Nt(t);
      Gr(function() {
        var o = !1;
        try {
          o = Yr();
        } finally {
          if (Ge = !1, t = r - Nt(), !Jr())
            return;
          o ? n.run(1e3) : t > 0 ? n.run(t) : n.start();
        }
      });
    }
  }, e.prototype.schedule = function() {
    this.stop(), this.run();
  }, e.prototype.observe = function() {
    var t = this, n = function() {
      return t.observer && t.observer.observe(document.body, Zr);
    };
    document.body ? n() : be.addEventListener("DOMContentLoaded", n);
  }, e.prototype.start = function() {
    var t = this;
    this.stopped && (this.stopped = !1, this.observer = new MutationObserver(this.listener), this.observe(), _t.forEach(function(n) {
      return be.addEventListener(n, t.listener, !0);
    }));
  }, e.prototype.stop = function() {
    var t = this;
    this.stopped || (this.observer && this.observer.disconnect(), _t.forEach(function(n) {
      return be.removeEventListener(n, t.listener, !0);
    }), this.stopped = !0);
  }, e;
})(), ot = new eo(), Wt = function(e) {
  !Le && e > 0 && ot.start(), Le += e, !Le && ot.stop();
}, to = function(e) {
  return !mt(e) && !Hr(e) && getComputedStyle(e).display === "inline";
}, no = (function() {
  function e(t, n) {
    this.target = t, this.observedBox = n || xe.CONTENT_BOX, this.lastReportedSize = {
      inlineSize: 0,
      blockSize: 0
    };
  }
  return e.prototype.isActive = function() {
    var t = mn(this.target, this.observedBox, !0);
    return to(this.target) && (this.lastReportedSize = t), this.lastReportedSize.inlineSize !== t.inlineSize || this.lastReportedSize.blockSize !== t.blockSize;
  }, e;
})(), ro = /* @__PURE__ */ (function() {
  function e(t, n) {
    this.activeTargets = [], this.skippedTargets = [], this.observationTargets = [], this.observer = t, this.callback = n;
  }
  return e;
})(), Ae = /* @__PURE__ */ new WeakMap(), It = function(e, t) {
  for (var n = 0; n < e.length; n += 1)
    if (e[n].target === t)
      return n;
  return -1;
}, Me = (function() {
  function e() {
  }
  return e.connect = function(t, n) {
    var r = new ro(t, n);
    Ae.set(t, r);
  }, e.observe = function(t, n, r) {
    var o = Ae.get(t), i = o.observationTargets.length === 0;
    It(o.observationTargets, n) < 0 && (i && oe.push(o), o.observationTargets.push(new no(n, r && r.box)), Wt(1), ot.schedule());
  }, e.unobserve = function(t, n) {
    var r = Ae.get(t), o = It(r.observationTargets, n), i = r.observationTargets.length === 1;
    o >= 0 && (i && oe.splice(oe.indexOf(r), 1), r.observationTargets.splice(o, 1), Wt(-1));
  }, e.disconnect = function(t) {
    var n = this, r = Ae.get(t);
    r.observationTargets.slice().forEach(function(o) {
      return n.unobserve(t, o.target);
    }), r.activeTargets.splice(0, r.activeTargets.length);
  }, e;
})(), $t = (function() {
  function e(t) {
    if (arguments.length === 0)
      throw new TypeError("Failed to construct 'ResizeObserver': 1 argument required, but only 0 present.");
    if (typeof t != "function")
      throw new TypeError("Failed to construct 'ResizeObserver': The callback provided as parameter 1 is not a function.");
    Me.connect(this, t);
  }
  return e.prototype.observe = function(t, n) {
    if (arguments.length === 0)
      throw new TypeError("Failed to execute 'observe' on 'ResizeObserver': 1 argument required, but only 0 present.");
    if (!Pt(t))
      throw new TypeError("Failed to execute 'observe' on 'ResizeObserver': parameter 1 is not of type 'Element");
    Me.observe(this, t, n);
  }, e.prototype.unobserve = function(t) {
    if (arguments.length === 0)
      throw new TypeError("Failed to execute 'unobserve' on 'ResizeObserver': 1 argument required, but only 0 present.");
    if (!Pt(t))
      throw new TypeError("Failed to execute 'unobserve' on 'ResizeObserver': parameter 1 is not of type 'Element");
    Me.unobserve(this, t);
  }, e.prototype.disconnect = function() {
    Me.disconnect(this);
  }, e.toString = function() {
    return "function ResizeObserver () { [polyfill code] }";
  }, e;
})();
const Ne = /* @__PURE__ */ new WeakMap();
function oo(e) {
  var t;
  for (let n = 0, r = e.length; n < r; ++n) {
    const o = e[n], i = Ne.get(o.target);
    if (typeof i == "function") {
      const { inlineSize: c, blockSize: s } = ((t = o.borderBoxSize) == null ? void 0 : t[0]) ?? {}, { offsetWidth: a, offsetHeight: f } = o.target;
      i(
        Object.assign(o, {
          offsetWidth: a,
          offsetHeight: f,
          width: c ?? a,
          height: s ?? f
        })
      );
    }
  }
}
const gn = new (F && window.ResizeObserver || $t)(
  oo
);
function Ee(e, t) {
  Ne.set(e, t), gn.observe(e);
}
function it(e) {
  Ne.has(e) && (gn.unobserve(e), Ne.delete(e));
}
function Lo(e = {}) {
  let t = B;
  const n = j(
    () => E(e.target),
    (o) => {
      t(), !(!o || typeof e.onResize != "function") && (Ee(o, e.onResize), t = () => {
        it(o), t = B;
      });
    },
    { immediate: !0 }
  ), r = () => {
    n(), t();
  };
  return Re() && U(r), {
    /**
     * @deprecated Will be removed in next major version, please directly use `observeResize` from imports.
     */
    observeResize: Ee,
    /**
     * @deprecated Will be removed in next major version, please directly use `unobserveResize` from imports.
     */
    unobserveResize: it,
    unobserve: r
  };
}
const bn = C(!1), io = W(() => bn.value), pe = "__rtl_observer__";
function so() {
  return F && !document.querySelector(`#${pe}`) && co(), { isRtl: io };
}
function co() {
  if (!F) return;
  const e = document.createElement("style"), t = `#${pe} { width: 1px } html.rtl #${pe}, html[dir='rtl'] #${pe} { width: 2px }`;
  e.textContent = t, document.head.appendChild(e);
  const n = document.createElement("div");
  n.id = pe, n.role = "none", n.style.cssText = "position: fixed; top -10px; left: -10px; height: 1px; visibility: hidden;", Ee(n, () => {
    F && (bn.value = document.documentElement.classList.contains("rtl") || document.documentElement.getAttribute("dir") === "rtl");
  }), document.body.appendChild(n);
}
const Po = Object.freeze([
  "top",
  "top-start",
  "top-end",
  "bottom",
  "bottom-start",
  "bottom-end",
  "left",
  "left-start",
  "left-end",
  "right",
  "right-start",
  "right-end"
]);
function Bo(e) {
  const { transfer: t, wrapper: n, isDrop: r = !1 } = e, o = e.reference ?? Ve(null), i = e.popper ?? Ve(null), c = e.arrow ?? Ve(null), s = C(E(e.placement)), a = C(""), f = e.isRtl ?? so().isRtl;
  n && gt(() => {
    const p = E(n), g = E(i);
    p && (p.__transferElement = g);
  }), t != null && (j(
    () => E(t),
    (p) => {
      h(p), v();
    }
  ), h(E(t))), ct(() => {
    requestAnimationFrame(() => {
      gt(u);
    });
  });
  let l;
  async function u() {
    if (!F) return;
    l == null || l();
    const p = E(o), g = E(i);
    if (!p || !g) return;
    const m = E(c), d = [Br()];
    if (r && d.push({
      name: "origin",
      fn({ placement: R, elements: A }) {
        const k = b(R);
        return k && (A.floating.style.transformOrigin = k), {};
      }
    }), e.offset) {
      let R = E(e.offset);
      Array.isArray(R) && (R = {
        mainAxis: R[1],
        crossAxis: R[0]
      }), d.push(Lr(R));
    }
    if (e.shift) {
      let R = E(e.shift);
      typeof R == "boolean" && (R = {}), d.push(Pr(R));
    }
    m && d.push(Fr({ element: m })), d.push(Dr({ strategy: "escaped" }));
    const w = E(f) || !1, O = {
      middleware: d,
      placement: E(e.placement),
      platform: {
        ...ln,
        isRTL: async () => w
      }
    }, T = async () => {
      const {
        x: R,
        y: A,
        placement: k,
        strategy: L,
        middlewareData: P
      } = await _r(p, g, O);
      if (E(o) !== p) {
        E(i) === g && Object.assign(g.style, {
          position: "",
          top: "",
          left: ""
        });
        return;
      }
      const $ = {
        position: L,
        top: `${A}px`,
        left: `${R}px`
      };
      if (m)
        if (P.arrow) {
          const { x: N, y } = P.arrow;
          Object.assign(m.style, {
            top: y != null ? `${y}px` : "",
            left: N != null ? `${N}px` : ""
          });
        } else
          Object.assign(m.style, { top: "", left: "" });
      Object.assign(g.style, $), g.dataset.popperPlacement = k, s.value = k;
    };
    e.autoUpdate && (l = Mr(p, g, T)), await T();
  }
  const v = () => new Promise((p) => {
    requestAnimationFrame(() => {
      u().then(p);
    });
  });
  function h(p) {
    a.value = typeof p == "boolean" ? p ? "body" : "" : p;
  }
  function b(p) {
    if (p !== "left" && p !== "right") {
      const [g, m] = p.split("-");
      return g === "bottom" || g !== "top" && m === "start" ? "center top" : "center bottom";
    }
  }
  return {
    wrapper: n,
    reference: o,
    popper: i,
    placement: s,
    transferTo: a,
    updatePopper: v
  };
}
const Ht = "__theme_style__", Pe = "__theme_observer__", qt = /* @__PURE__ */ new WeakMap();
function Do(e) {
  const t = C();
  if (F) {
    e = e || document.body;
    let n = qt.get(e);
    n || (n = uo(t), e.appendChild(n), qt.set(e, n));
  }
  return { theme: t };
}
const We = Je(/* @__PURE__ */ new Map()), st = /* @__PURE__ */ new Map();
j(We, () => {
  if (!F) return;
  st.clear();
  const e = document.head.querySelector(`#${Ht}`);
  e && document.head.removeChild(e);
  const t = document.createElement("style");
  let n = `.${Pe} { width: 1px }`, r = 1;
  for (const [o, [i, c]] of We.entries())
    n += ` html.${i} .${Pe}, .${c} .${Pe} { width: ${++r}px }`, st.set(r, o);
  t.textContent = n, t.id = Ht, document.head.appendChild(t);
});
function ao(e) {
  for (const t of e) {
    const {
      name: n,
      rootClass: r = n,
      varsClass: o = `vxp-theme-vars-${r}`
    } = typeof t == "string" ? { name: t } : t;
    We.set(n, [r, o]);
  }
}
function Fo(e) {
  We.clear(), ao(e);
}
let lo = 0;
function fo() {
  return `__theme_observer_${lo++}__`;
}
function uo(e) {
  if (!F) return;
  const t = document.createElement("div");
  return t.id = fo(), t.className = Pe, t.role = "none", t.style.cssText = "position: fixed; top -10px; left: -10px; height: 1px; visibility: hidden;", Ee(t, (n) => {
    var o;
    if (!F) return;
    const r = ((o = n.borderBoxSize) == null ? void 0 : o[0].inlineSize) ?? n.contentRect.width;
    e.value = st.get(r);
  }), t;
}
function ho() {
  const e = {};
  return U(() => {
    Object.keys(e).forEach((t) => {
      clearTimeout(e[t]);
    });
  }), { timer: e };
}
function mo() {
  const e = {};
  return U(() => {
    Object.keys(e).forEach((t) => {
      clearInterval(e[t]);
    });
  }), { timer: e };
}
function vo() {
  const e = {};
  return U(() => {
    Object.keys(e).forEach((t) => {
      cancelAnimationFrame(e[t]);
    });
  }), { timer: e };
}
function _o() {
  const { timer: e } = ho(), { timer: t } = mo(), { timer: n } = vo();
  return { timeout: e, interval: t, raf: n };
}
function No(e, t, n = 100) {
  let r;
  function o() {
    e.value === "hover" && (clearTimeout(r), r = setTimeout(() => {
      t.value = !0;
    }, n));
  }
  function i() {
    e.value === "hover" && (clearTimeout(r), r = setTimeout(() => {
      t.value = !1;
    }, n));
  }
  function c() {
    e.value === "click" && (t.value = !t.value);
  }
  return {
    handleTriggerEnter: o,
    handleTriggerLeave: i,
    handleTriggerClick: c
  };
}
function Wo(e) {
  const {
    items: t,
    itemSize: n,
    itemFixed: r,
    idKey: o,
    defaultKeyAt: i,
    bufferSize: c = C(5),
    wrapper: s = C(null),
    autoResize: a = !0
  } = e, f = W(() => {
    const y = E(t), x = y.length, S = E(o), M = /* @__PURE__ */ new Map();
    for (let z = 0; z < x; ++z)
      M.set(y[z][S], z);
    return M;
  }), l = C(0), u = /* @__PURE__ */ new Map(), v = W(() => {
    const y = E(t), x = y.length, S = E(o), M = Cn(x, E(n));
    for (let z = 0; z < x; ++z) {
      const D = y[z][S], _ = u.get(D);
      _ && M.add(z, _);
    }
    return M;
  }), h = C(0), b = C(0), p = W(() => Math.max(
    v.value.boundIndex(b.value) - Math.max(E(c), 0),
    0
  )), g = W(() => {
    if (!l.value || l.value < 0) return [];
    const y = E(t), x = Math.min(
      v.value.boundIndex(b.value + l.value) + 1 + Math.max(E(c), 0),
      y.length
    );
    return y.slice(p.value, x);
  }), m = W(() => {
    h.value;
    const y = v.value.sum(), x = E(r);
    return {
      height: x ? `${y}px` : void 0,
      minHeight: x ? void 0 : `${y}px`,
      boxSizing: "content-box"
    };
  }), d = W(() => (h.value, {
    transform: `translate3d(0, ${v.value.sum(p.value)}px, 0)`
  }));
  if (j(
    () => E(s),
    (y) => {
      y && Be(i) && requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          L(i);
        });
      });
    },
    { immediate: !0, flush: "post" }
  ), a) {
    let y = B;
    const x = j(
      () => [E(s), a],
      ([S]) => {
        y(), S && (Ee(S, T), y = () => {
          it(S), y = B;
        }, Be(i) && L(i));
      },
      { immediate: !0, flush: "post" }
    );
    U(() => {
      x(), y();
    });
  }
  function w() {
    const y = E(s);
    y && (b.value = y.scrollTop);
  }
  function O() {
    Fn(w);
  }
  function T(y) {
    xt(y.target) || y.contentRect.height === l.value || (l.value = y.contentRect.height);
  }
  function R(y, x) {
    var vt, pt;
    if (E(r)) return;
    const S = f.value.get(y), M = v.value, z = M.get(S), D = ((pt = (vt = x.borderBoxSize) == null ? void 0 : vt[0]) == null ? void 0 : pt.blockSize) ?? x.contentRect.height;
    if (D === z) return;
    if (xt(x.target)) {
      z && (M.add(S, -z), h.value++);
      return;
    }
    const _ = D - E(n), ee = D - z;
    if (_ ? u.set(y, _) : u.delete(y), !ee) return;
    M.add(S, ee), h.value++;
    const te = E(s);
    if (te) {
      const wn = M.sum(S);
      te.scrollTop > wn && te.scrollBy(0, ee), b.value = te.scrollTop;
    }
  }
  function A(y, x) {
    const S = E(s);
    S && S.scrollTo({
      behavior: x,
      top: y,
      left: 0
    });
  }
  function k(y, x) {
    const S = E(s);
    S && S.scrollBy({
      behavior: x,
      top: y,
      left: 0
    });
  }
  function L(y, x) {
    const S = f.value.get(y);
    S != null && P(S, x);
  }
  function P(y, x) {
    const S = E(s);
    S && S.scrollTo({
      behavior: x,
      top: v.value.sum(y),
      left: 0
    });
  }
  function $(y, x) {
    const S = E(s);
    if (!S) return;
    const M = v.value, z = S.scrollTop;
    if (M.sum(y) < z) {
      P(y, x);
      return;
    }
    const _ = S.offsetHeight, ee = z + _, te = M.sum(y + 1);
    te > ee && A(te - _, x);
  }
  function N(y, x) {
    const S = f.value.get(y);
    S != null && $(S, x);
  }
  return {
    wrapper: s,
    indexMap: f,
    heightTree: v,
    startIndex: p,
    scrollOffset: b,
    visibleItems: g,
    listStyle: m,
    itemsStyle: d,
    handleScroll: O,
    handleResize: T,
    handleItemResize: R,
    scrollTo: A,
    scrollBy: k,
    scrollToKey: L,
    scrollToIndex: P,
    ensureIndexInView: $,
    ensureKeyInView: N
  };
}
export {
  Ut as CLICK_OUTSIDE,
  ao as addActiveThemes,
  wo as createSlotRender,
  yo as flatVNodes,
  xt as isHiddenElement,
  Ee as observeResize,
  Ro as pickToRefs,
  Po as placementWhileList,
  xo as proxyExposed,
  Gt as queryOutsideHiddenElement,
  Fo as setActiveThemes,
  it as unobserveResize,
  Te as unrefElement,
  go as useClickOutside,
  bo as useDisplay,
  So as useFullScreen,
  Oo as useHover,
  To as useIntersection,
  K as useListener,
  zo as useManualRef,
  Co as useModifier,
  ko as useMounted,
  Ao as useMoving,
  Mo as usePointer,
  Bo as usePopper,
  vo as useRaf,
  Lo as useResize,
  so as useRtl,
  mo as useSetInterval,
  ho as useSetTimeout,
  Do as useTheme,
  _o as useTimerRecord,
  No as useTriggerHandler,
  Wo as useVirtual,
  Eo as watchPauseable
};
